typegen --watch (#12210)
remix-run/react-router: docs/how-to/setting-up-type-safety.md

---
title: Setting up type safety
---

To know more about how type safety works in React Router, check out our [dedicated explanation](../explanation//type-safety.md).

# Setting up type safety

> React Router generates types for into a `.react-router/` directory at the root of your app.

Reviewer: ```suggestion
React Router generates types into a `.react-router/` directory at the root of your app.
```

This directory is fully managed by React Router and is derived from your route config (`app/routes.ts` by default), so it should be gitignore'd.

👉 **Add `.react-router/` to `.gitignore`**

```txt
.react-router/
```

> You should also ensure that generated types are always present before running typechecking,
> especially for running typechecking in CI.

Reviewer: Small tweak, I think just state the instruction instead of "you should" in general sounds good for this sort of a doc

```suggestion
Make sure generated types are always present before running type checking,
especially when running type checking in CI.
```

👉 **Add `react-router typegen` to your `typecheck` command in `package.json`**

```json
{
  "scripts": {
    "typecheck": "react-router typegen && tsc"
  }
}
```

To get TypeScript to use those generated types, you'll need to add them to `include` in `tsconfig.json`.
And to be able to import them as if they files next to your route modules, you'll also need to [configure `rootDirs`](https://www.typescriptlang.org/tsconfig/#rootDirs).

👉 **Configure `tsconfig.json` for generated types**

```json
{
  "include": [".react-router/types/**/*"],
  "compilerOptions": {
    "rootDirs": [".", "./.react-router/types"]
  }
}
```

## Automatic typegen in VSCode (optional)

> If you'd rather not need to remember to run `react-router typegen --watch` every time you start working on your app, you can use [VSCode Tasks](https://code.visualstudio.com/docs/editor/tasks) to automate this.

Reviewer: This is the first time the watch flag has been mentioned. Should this command be referenced above? I imagine whether you're using vscode or not you'll want to add this script
